src/hw5/COMP232LinkedBinarySearchTree.java

```diff
@@ -170,8 +194,37 @@ private void addNodeToSubTree(BTNode<K, V> subTreeRoot,
 	 * {@inheritDoc}
 	 */
 	public V remove(K key) {
-		// Intentionally not implemented - see homework assignment.
-		throw new UnsupportedOperationException("Not yet implemented");
+		BTNode<K, V> cur = root;
+		while (true) {
+			if (cur == null) {
+				return null;
+			} else if (cur.key == key) {
+				if (cur.parent.left == cur) {
+					cur.parent.left = null;
+				} else if (cur.parent.right == cur) {
+					cur.parent.right = null;
+				}
+
+				//"If the node has one or more descendants, its rightmost leaf takes its place in the tree"
+				boolean has_right = false;
+				if (cur.right != null) {
+					has_right = true;
+					addNodeToSubTree(cur.parent, cur.right);
+				}
+				if (cur.left != null) {
+					if (!has_right) {
+						addNodeToSubTree(cur.parent, cur.left);
+					} else {
+						addNodeToSubTree(root, cur.left);
+					}
+				}
+				return cur.value;
```

Comment:
We should also decrement the size of the tree, this causes a couple of errors in the test cases I provided in No11Test.java